imgui-SFML.cpp

```diff
@@ -666,17 +765,178 @@ void Update(sf::Window& window, sf::RenderTarget& target, sf::Time dt) {
+                            wc->joystickId = getConnectedJoystickId();
+                        }
+                        break;
+                    default:
+                        break;
+                    }
+                }
+
+                switch (event.type) {
+                case sf::Event::LostFocus: {
+                    io.AddFocusEvent(false);
+                    wc->windowHasFocus = false;
+                } break;
+                case sf::Event::GainedFocus:
+                    io.AddFocusEvent(true);
+                    wc->windowHasFocus = true;
+                    break;
+                case sf::Event::MouseEntered:
+                    {
+                        sf::Vector2i mousePos = sf::Mouse::getPosition();
+                        io.AddMousePosEvent(mousePos.x, mousePos.y);
+                    }
+                    wc->mouseHovered = true;
+                    break;
+                case sf::Event::MouseLeft:
+                    io.AddMousePosEvent(-FLT_MAX, -FLT_MAX);
+                    wc->mouseHovered = false;
+                    break;
+                default:
+                    break;
+                }
+            }
+        }
+    }
+    
+    ImGuiPlatformIO& platform_io = ImGui::GetPlatformIO();
+    const ImVec2 mousePosPrev = io.MousePos;
+    ImGuiID mouseViewportID = 0;
+
+    for (auto* viewport : platform_io.Viewports) {
+        WindowContext* wc = (WindowContext*)viewport->PlatformUserData;
+        sf::Window* window = (sf::Window*)viewport->PlatformHandle;
+
+        if (window->hasFocus()) {
+            if (io.WantSetMousePos) {
+                sf::Mouse::setPosition({static_cast<int>(mousePosPrev.x - viewport->Pos.x),
+                                        static_cast<int>(mousePosPrev.y - viewport->Pos.y)},
+                                       *window);
+            }
+            sf::Vector2i mousePos;
+            if (io.ConfigFlags & ImGuiConfigFlags_ViewportsEnable)
+                mousePos = sf::Mouse::getPosition();
+            else
+                mousePos = sf::Mouse::getPosition(*window);
+            
+            io.AddMousePosEvent(mousePos.x, mousePos.y);
+
+            for (unsigned int i = 0; i < 3; i++) {
+                io.MouseDown[i] = wc->touchDown[i] || sf::Touch::isDown(i) || wc->mousePressed[i] ||
+                                  sf::Mouse::isButtonPressed((sf::Mouse::Button)i);
+                wc->mousePressed[i] = false;
+                wc->touchDown[i] = false;
+            }
+        }
+
+        const bool windowNoInput = (viewport->Flags & ImGuiViewportFlags_NoInputs) != 0;
+
+        if (wc->mouseHovered && !windowNoInput) mouseViewportID = viewport->ID;
+    }
+
+    if (io.BackendFlags & ImGuiBackendFlags_HasMouseHoveredViewport)
+        io.AddMouseViewportEvent(mouseViewportID);
+
+#else
     if (s_currWindowCtx->windowHasFocus) {
```

Comment:
Why doesn't this code need to be called when viewports are on?

Reply:
Because it is called while iterating through viewports. In ImGui main window is also considered a viewport and it is the first item in `ImGuiPlatformIO::Viewports` vector.